### DISCUSSION TITLE
buy nft with smart contrtact solidity using fulFillBasicOrder

### DISCUSSION THREAD

**Discussion by jecombe** [ProjectOpenSea/seaport#964] [Q&A]
i've create a smart contract who buy nft, and i'm using opensea-js to obtain basicOrderParam with function getOrder. But i think i'm not good..
this my contract: `contract OpenSea {
    AdditionalRecipient[] additionalRecipients;
    event Received(address, uint256);

    constructor() {}

   

    function buyNft() public {
        AdditionalRecipient memory seaport = AdditionalRecipient({
            amount: 25000000000000000,
            recipient: payable(0x0000a26b00c1F0DF003000390027140000fAa719)
        });

        additionalRecipients.push(seaport);

        BasicOrderParameters memory basicOrder = BasicOrderParameters({
            considerationToken: 0x0000000000000000000000000000000000000000,
            considerationIdentifier: 0,
            considerationAmount: 9750000000000000,
            offerer: payable(0x6Cc544212340Ae755454Cc94f19D3Cec32Bf8402),
            zone: 0x0000000000000000000000000000000000000000,
            offerToken: 0x317a8Fe0f1C7102e7674aB231441E485c64c178A,
            offerIdentifier: 532583,
            offerAmount: 1, 
            basicOrderType: BasicOrderType.ETH_TO_ERC721_FULL_OPEN,
            startTime: 1675398225,
            endTime: 1677817425,
            zoneHash: 0x0000000000000000000000000000000000000000000000000000000000000000,
            salt: 0x360c6ebe0000000000000000000000000000000000000000a53bb8e9ff2fb762,
            offererConduitKey: 0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000,
            fulfillerConduitKey: 0x0000007b02230091a7ed01230072f7006a004d60a8d4e71d599b8104250f0000,
            totalOriginalAdditionalRecipients: 1,
            additionalRecipients: additionalRecipients,
            signature: "0x3638a108891179952dd78386aa7d08a4a314ea2ef0d5da6454445f8a03bbfcb67d2a2aaadc21015e9d309472e89b6561e03948ea81399b07d4ae820f9d888c561b"
        });

        bool status = SeaportInterface(
            0x00000000006c3852cbEf3e08E8dF289169EdE581
        ).fulfillBasicOrder(basicOrder);
    }

    receive() external payable {
        emit Received(msg.sender, msg.value);
    }
}`
does anyone have a working example? thank's

**Comment by mustafa222lk** [comment]
you call fulfill and then it gets upgraded to a basicorder.

**Comment by ryanio** [comment]
what is the error you're running into? you don't need to do this in a contract, you can send a tx to fulfillBasicOrder directly

**Reply by DistractionBoy** [reply]
@ryanio is there an example of taking the results of Fulfill Listing (.../v2/listings/fulfillment_data) and then calling fulfillBasicOrder from a connected Seaport contract like:
```
          const browserProvider = new ethers.BrowserProvider(
            window.ethereum,
            chain
          );
          const signer = await browserProvider.getSigner(userWalletAddress);
          const contract = new Contract(
            order["fulfillment_data"].transaction.to,
            seaportAbi,
            signer
          );
          const connectedContract = contract.connect(signer);
          contract.fulfillBasicOrder({
            ...order["fulfillment_data"].transaction,
          });
```

something like that... how do we invoke the returned function to fulfill the order at order["fulfillment_data"].transaction.function ?

(in a react/Nextjs web client)

**Reply by ryanio** [reply]
It would be easiest to just use opensea-js, it will take care of formatting the purchase tx for you.

**Reply by DistractionBoy** [reply]
thank you so much @ryanio for those reading this with the same / similar mission...

using the opensea-js package did help me with resolving types and stuff, but having to instantiate the sdk by chain would mean i'd have three instances floating around (for my website) and i really wanted something a little more lightweight. after hopping in the discord, thenerdassassin sent me this link https://github.com/montasaurus/bidding-bot-demo/blob/main/src/fulfillOffer.ts which showed me what i needed for my use case which is client side. 

resulting code (before cleanup):

```js
const getFulfillmentTransactionData = async (
  url: string,
  {
    arg,
  }: {
    arg: {
      listing: {
        chain: string;
        hash: string;
        protocol_address: string;
      };
      fulfiller: { address: `0x${string}` | undefined };
    };
  }
) => {
  return await fetch(url, {
    method: "POST",
    headers: {
      accept: "application/json",
      "content-type": "application/json",
    },
    body: JSON.stringify(arg),
  }).then((res) => res.json());
};
```

```js
  const {
    trigger,
    isMutating,
    data: order,
    error: orderError,
  } = useSWRMutation(
    `/api/opensea/fulfillListing`,
    getFulfillmentTransactionData,
    {
      onSuccess: async (order) => {
        try {
          const orderTransaction = order.fulfillment_data.transaction;
          const seaportInterface = new ethers.Interface(seaportAbi);
          const fragment = seaportInterface.getFunction(
            orderTransaction.function
          );
          if (fragment === null) {
            throw new Error("fulfill_order function not present");
          } else {
            const encodedData = seaportInterface.encodeFunctionData(
              fragment,
              Object.values(orderTransaction.input_data)
            );
            const txReq: TransactionRequest = {
              to: orderTransaction.to,
              from: userWalletAddress,
              value: orderTransaction.value,
              data: encodedData,
            };
            setTx(txReq);
          }
        } catch (e: any) {
          toast(e.message);
        }
      },
    }
  );
```
```js
<Button
  disabled={isMutating}
  onClick={async () => {
    try {
      if (
        window.ethereum === null ||
        window.ethereum === undefined
      ) {
        throw new Error("wallet not connected");
      }
      const browserProvider = new ethers.BrowserProvider(
        window.ethereum,
        chain
      );
      const signer: ethers.Signer =
        await browserProvider.getSigner(userWalletAddress);

      sign(signer, "OptiChads Floor Buy").then(
        async (signature) => {
          const result = verifyMessage(
            "OptiChads Floor Buy",
            signature as string
          );
          if (result !== userWalletAddress) {
            throw new Error(
              "Singature failed, are you who you say you are bro?"
            );
          }
          if (tx === undefined) {
            throw new Error(
              "Failed to get transaction details. Try refreshing the page and try again."
            );
          }
          const receipt: ethers.TransactionResponse =
            await signer.sendTransaction(tx);
          toast(
            `Purchase successful. Tx hash: ${receipt.hash}`
          );
        }
      );
    } catch (e: any) {
      toast(e.mesage);
    }
  }}
>
  BUY NOW BRO
</Button>
  
```
